### PR TITLE
[ Misc ] Refactor w8a8 to use `process_weights_after_load` (Simplify Weight Loading)

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* simplify weight loading logic by converting `w8a8` loading to use `process_weights_after_load` to convert the N logical per-tensor scales into channel-wise scales. This avoids having to pass `logical_widths` to `linear.py`, which reduces complexity in that function + combines int8 and fp8 logic together
* rename to `is_per_tensor_scale` to be more explicit about the special casing

FOLLOW UP TO: https://github.com/vllm-project/vllm/pull/5928

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/fp8.py [hunk 1]
```diff
@@ -112,12 +112,10 @@ def _create_scale_param(
...
         layer.register_parameter(scale_name, scale)
-        set_weight_attrs(
-            scale, {
-                **extra_weight_attrs,
-                "fp8_scales_shard_indexer":
-                self.scales_shard_indexer,
-            })
+        set_weight_attrs(scale, {
+            **extra_weight_attrs,
+            "is_per_tensor_scale": True,
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/fp8.py:117]: I feel per-tensor scale is more commonly used than per-channel. Do you think it'd be more concise if we use "is_per_channel_scale" and only set the value for w8a8?

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fp8.py:117]: * The `per-tensor-scase` case is the special case for weight loading, because we need to handle loading 1 scale into an array of N scales since we have fusion for QKV and MLP
* We actually don't have a problem for channelwise, since we just make a buffer that is `sum(partition_sizes)` and load with the same logic we use for weights
* So we need an attribute for `is_per_tensor_scale` that the weight loader can branch off to do the indexing properly

File: tests/quantization/test_compressed_tensors.py [hunk 2]
```diff
@@ -34,17 +36,21 @@ def test_compressed_tensors_w8a8_static_setup(vllm_runner, model_args):
...
+                         else torch.float8_e4m3fn)
+
+        assert qkv_proj.weight.dtype is expected_type
+        assert o_proj.weight.dtype is expected_type
+        assert gate_up_proj.weight.dtype is expected_type
 
         if qkv_proj.scheme.strategy == "tensor":
-            assert qkv_proj.weight_scale.shard_splitter is not None
-            assert qkv_proj.weight_scale.logical_widths is not None
+            breakpoint()
```
Comment (robertgshaw2-neuralmagic) [inline tests/quantization/test_compressed_tensors.py:51]: fixing

File: vllm/model_executor/layers/quantization/fp8.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fp8.py:117]: Basically, we used to branch on `fp8_weight_shard_indexer` or something like that ... this is the replacement + the functionality is shared across both int8 and fp8 for this case since its the same logic

Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:117]: That's a fair point. I guess I was just a bit confused about the naming is_per_tensor_scale being used as a special case for fused layers.

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fp8.py:117]: It’s hard to pick a name. I’ll think of something else

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py [hunk 3]
```diff
@@ -15,70 +15,63 @@ class CompressedTensorsW8A8(CompressedTensorsScheme):
...
-        offset = sum(logical_widths[:shard_id])
-        size = logical_widths[shard_id]
-        # update loaded weight with copies for broadcast.
-        loaded_weight = loaded_weight.repeat(size)
-        return param[offset:offset + size], loaded_weight
+    # Cutlass kernels support only per-tensor and per-channel cases.
+    # So if we have a fused module (QKV, MLP) with per tensor scales (thus N
+    # scales being passed to the kernel), we convert to the per-channel case.
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        if (self.strategy == QuantizationStrategy.TENSOR
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py:22]: nit: could just move this to the per tensor scheme

**Review by dsikka — APPROVED** [review]
LGTM

**Comment by robertgshaw2-neuralmagic** [comment]
@comaniac updated per your request (+ merged with the phi PR)

**Review by comaniac — APPROVED** [review]
LGTM. Thanks

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py (hunk 3 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py:22]: strategy refers to the weight quantization here
